<a href="https://colab.research.google.com/github/DanielRaiicHu/TelecomX/blob/main/TelecomX.ipynb?forceRefresh=true">
  <img src="https://drive.google.com/uc?export=view&id=191Woxwut0iaIL2hCcCIeXRSB6oba_88H" alt="Abrir en Colab" width="120"/>
</a>

## **Archivo del proyecto**

https://raw.githubusercontent.com/alura-cursos/challenge2-data-science-LATAM/refs/heads/main/TelecomX_Data.json

Exploración básica
```
df.shape           # Dimensiones del DataFrame
df.columns         # Lista de nombres de columnas
df.head()          # Primeras 5 filas
df.info()          # Resumen de tipos de datos y valores nulos
df.describe()      # Estadísticas numéricas (si hay números)
df.dtypes          # Tipo de dato por columna
df.isnull().sum()  # Conteo de valores nulos por columna
```



## **Extracción de datos**

In [58]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

### **Leyendo los datos**

In [59]:
url = 'https://raw.githubusercontent.com/alura-cursos/challenge2-data-science-LATAM/refs/heads/main/TelecomX_Data.json'

In [60]:
df = pd.read_json(url)

In [61]:
df.head()

customerID Churn                                           customer  \
0  0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1  0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2  0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3  0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4  0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   

                                             phone  \
0   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   

                                            internet  \
0  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1  {'InternetService': 'DSL', 'OnlineSecurity': '...   
2  {'InternetService': 'Fiber optic', 'OnlineSecu...   
3  {'InternetService': 'Fiber optic', 'OnlineSecu...   
4  {'InternetService': 'Fiber optic', 'OnlineSecu...   

                                             account  
0  {'Contract': 'One year', 'PaperlessBilling': '...  
1  {'Contract': 'Month-to-month', 'PaperlessBilli...  
2  {'Contract': 'Month-to-month', 'PaperlessBilli...  
3  {'Contract': 'Month-to-month', 'PaperlessBilli...  
4  {'Contract': 'Month-to-month', 'PaperlessBilli...

### **Normalizando el dataframe**

Debemos normalizar el dataframe ya que presenta información anidada.
Sobreescribiremos el dataframe.

In [62]:
# Normalizo cada una de las columnas que contienen diccionarios
df_customer = pd.json_normalize(df['customer'])
df_phone = pd.json_normalize(df['phone'])
df_internet = pd.json_normalize(df['internet'])
df_account = pd.json_normalize(df['account'])

# Concateno las columnas con las columnas normalizadas en el dataframe
df = pd.concat([df[['customerID', 'Churn']], df_customer, df_phone, df_internet, df_account], axis=1)
df.head()

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
0  0002-ORFBO    No  Female              0     Yes        Yes       9   
1  0003-MKNFE    No    Male              0      No         No       9   
2  0004-TLHLJ   Yes    Male              0      No         No       4   
3  0011-IGKFF   Yes    Male              1     Yes         No      13   
4  0013-EXCHZ   Yes  Female              1     Yes         No       3   

  PhoneService MultipleLines InternetService  ... OnlineBackup  \
0          Yes            No             DSL  ...          Yes   
1          Yes           Yes             DSL  ...           No   
2          Yes            No     Fiber optic  ...           No   
3          Yes            No     Fiber optic  ...          Yes   
4          Yes            No     Fiber optic  ...           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No         Yes         Yes              No        One year   
1               No          No          No             Yes  Month-to-month   
2              Yes          No          No              No  Month-to-month   
3              Yes          No         Yes             Yes  Month-to-month   
4               No         Yes         Yes              No  Month-to-month   

  PaperlessBilling     PaymentMethod Charges.Monthly  Charges.Total  
0              Yes      Mailed check            65.6          593.3  
1               No      Mailed check            59.9          542.4  
2              Yes  Electronic check            73.9         280.85  
3              Yes  Electronic check            98.0        1237.85  
4              Yes      Mailed check            83.9          267.4  

[5 rows x 21 columns]

In [155]:
df.sample(20)

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
3375  4686-UXDML    No  Female              0      No         No      21   
5704  7795-CFOCW    No    Male              0      No         No      45   
4373  5995-SNNEW    No    Male              1     Yes         No      23   
3158  4402-FTBXC    No    Male              0      No         No      54   
3343  4648-YPBTM    No    Male              0      No         No      53   
3340  4646-QZXTF    No  Female              0     Yes         No       7   
5196  7096-UCLNH    No    Male              0      No         No       1   
4708  6457-GIRWB   Yes    Male              0      No         No       1   
1482  2126-GSEGL    No  Female              0     Yes         No      70   
7001  9639-BUJXT    No    Male              0      No         No      63   
1414  2038-OEQZH    No    Male              0      No         No       4   
5934  8100-HZZLJ    No  Female              0      No        Yes      43   
5033  6885-PKOAM    No  Female              0     Yes         No      71   
4770  6537-OTKMY    No    Male              0      No         No      15   
1003  1403-LKLIK    No  Female              0     Yes        Yes      33   
3090  4304-XUMGI    No    Male              1     Yes        Yes      50   
4979  6813-GZQCG    No  Female              0     Yes        Yes      45   
3542  4883-KCPZJ    No  Female              0     Yes        Yes      22   
3186  4450-MDZFX    No    Male              0     Yes        Yes      60   
6072  8292-ITGYJ    No  Female              0     Yes        Yes      45   

     PhoneService     MultipleLines InternetService  ...         OnlineBackup  \
3375          Yes               Yes     Fiber optic  ...                  Yes   
5704           No  No phone service             DSL  ...                   No   
4373          Yes               Yes     Fiber optic  ...                  Yes   
3158          Yes                No     Fiber optic  ...                   No   
3343          Yes               Yes             DSL  ...                  Yes   
3340          Yes                No             DSL  ...                   No   
5196          Yes                No              No  ...  No internet service   
4708          Yes                No     Fiber optic  ...                   No   
1482          Yes               Yes             DSL  ...                  Yes   
7001          Yes                No              No  ...  No internet service   
1414          Yes                No     Fiber optic  ...                   No   
5934          Yes                No             DSL  ...                  Yes   
5033          Yes               Yes             DSL  ...                  Yes   
4770           No  No phone service             DSL  ...                   No   
1003          Yes                No              No  ...  No internet service   
3090          Yes               Yes             DSL  ...                  Yes   
4979          Yes               Yes              No  ...  No internet service   
3542          Yes               Yes              No  ...  No internet service   
3186          Yes               Yes     Fiber optic  ...                   No   
6072          Yes               Yes              No  ...  No internet service   

         DeviceProtection          TechSupport          StreamingTV  \
3375                   No                   No                  Yes   
5704                  Yes                  Yes                   No   
4373                  Yes                   No                  Yes   
3158                  Yes                   No                  Yes   
3343                  Yes                  Yes                   No   
3340                  Yes                  Yes                  Yes   
5196  No internet service  No internet service  No internet service   
4708                   No                   No                   No   
1482                  Yes                  Yes                  Yes 

## **Análisis de los datos**

In [93]:
# Dimensiones del Datafrane
df.shape

(7267, 21)

In [94]:
# Las columnas que tenemos en el dataframe
df.columns

Index(['customerID', 'Churn', 'gender', 'SeniorCitizen', 'Partner',
       'Dependents', 'tenure', 'PhoneService', 'MultipleLines',
       'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection',
       'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract',
       'PaperlessBilling', 'PaymentMethod', 'Charges.Monthly',
       'Charges.Total'],
      dtype='object')

In [158]:
# Para ver la información relevante y tipos de datos de las columnas
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7267 non-null   object 
 1   Churn             7267 non-null   object 
 2   gender            7267 non-null   object 
 3   SeniorCitizen     7267 non-null   int64  
 4   Partner           7267 non-null   object 
 5   Dependents        7267 non-null   object 
 6   tenure            7267 non-null   int64  
 7   PhoneService      7267 non-null   object 
 8   MultipleLines     7267 non-null   object 
 9   InternetService   7267 non-null   object 
 10  OnlineSecurity    7267 non-null   object 
 11  OnlineBackup      7267 non-null   object 
 12  DeviceProtection  7267 non-null   object 
 13  TechSupport       7267 non-null   object 
 14  StreamingTV       7267 non-null   object 
 15  StreamingMovies   7267 non-null   object 
 16  Contract          7267 non-null   object 


In [159]:
# Revisión de datos únicos para cada columna
for col in df.columns:
    if df[col].dtype == 'object':
        print(f"Valores únicos en '{col}':")
        print(df[col].unique())
        print("—" * 40)

Valores únicos en 'customerID':
['0002-ORFBO' '0003-MKNFE' '0004-TLHLJ' ... '9992-UJOEL' '9993-LHIEB'
 '9995-HOTOH']
————————————————————————————————————————
Valores únicos en 'Churn':
['No' 'Yes' '']
————————————————————————————————————————
Valores únicos en 'gender':
['Female' 'Male']
————————————————————————————————————————
Valores únicos en 'Partner':
['Yes' 'No']
————————————————————————————————————————
Valores únicos en 'Dependents':
['Yes' 'No']
————————————————————————————————————————
Valores únicos en 'PhoneService':
['Yes' 'No']
————————————————————————————————————————
Valores únicos en 'MultipleLines':
['No' 'Yes' 'No phone service']
————————————————————————————————————————
Valores únicos en 'InternetService':
['DSL' 'Fiber optic' 'No']
————————————————————————————————————————
Valores únicos en 'OnlineSecurity':
['No' 'Yes' 'No internet service']
————————————————————————————————————————
Valores únicos en 'OnlineBackup':
['Yes' 'No' 'No internet service']
————————————————————

In [126]:
# Analizando datos
df['Charges.Total'][0]

'593.3'

In [133]:
# Tipos de datos
type(df['SeniorCitizen'][0])

numpy.int64

In [160]:
# Ver las filas con el campo en blanco en la columna Churn
df[df['Churn'] == '']

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
30    0047-ZHDTW        Female              0      No         No      11   
75    0120-YZLQA          Male              0      No         No      71   
96    0154-QYHJU          Male              0      No         No      29   
98    0162-RZGMZ        Female              1      No         No       5   
175   0274-VVQOQ          Male              1     Yes         No      65   
...          ...   ...     ...            ...     ...        ...     ...   
7158  9840-GSRFX        Female              0      No         No      14   
7180  9872-RZQQB        Female              0     Yes         No      49   
7211  9920-GNDMB          Male              0      No         No       9   
7239  9955-RVWSC        Female              0     Yes        Yes      67   
7247  9966-VYRTZ        Female              0     Yes        Yes      31   

     PhoneService     MultipleLines InternetService  ...         OnlineBackup  \
30            Yes               Yes     Fiber optic  ...                   No   
75            Yes                No              No  ...  No internet service   
96            Yes                No             DSL  ...                  Yes   
98            Yes                No             DSL  ...                  Yes   
175           Yes               Yes     Fiber optic  ...                  Yes   
...           ...               ...             ...  ...                  ...   
7158          Yes               Yes             DSL  ...                  Yes   
7180           No  No phone service             DSL  ...                   No   
7211          Yes               Yes     Fiber optic  ...                   No   
7239          Yes                No              No  ...  No internet service   
7247          Yes                No              No  ...  No internet service   

         DeviceProtection          TechSupport          StreamingTV  \
30                     No                   No                   No   
75    No internet service  No internet service  No internet service   
96                     No                  Yes                   No   
98                     No                  Yes                   No   
175                   Yes                   No                  Yes   
...                   ...                  ...                  ...   
7158                   No                   No                   No   
7180                   No                   No                  Yes   
7211                   No                   No                   No   
7239  No internet service  No internet service  No internet service   
7247  No internet service  No internet service  No internet service   

          StreamingMovies        Contract PaperlessBilling  \
30                     No  Month-to-month              Yes   
75    No internet service        Two year              Yes   
96                     No        One year              Yes   
98                     No  Month-to-month               No   
175                   Yes        One year              Yes   
...                   ...             ...              ...   
7158                   No        One year              Yes   
7180                   No  Month-to-month               No   
7211                   No  Month-to-month              Yes   
7239  No internet service        Two year              Yes   
7247  No internet service  Month-to-month              Yes   

                  PaymentMethod Charges.Monthly  Charges.Total  
30    Bank transfer (automatic)           79.00          929.3  
75      Credit card (automatic)           19.90         1355.1  
96             Electronic check           58.75         1696.2  
98      Credit card (automatic)           59.90         287.85  
175   Bank transfer (automatic)          103.15        6792.45  
...                         ...             ...            ...  
7158               Mailed check           54.25          773.2  
71

In [161]:
# Cantidad de campos en blanco en la columna de Churn
(df['Churn'] == '').sum()

np.int64(224)

### **Análisis y decisiones sobre la limpieza de datos**

Tras revisar individualmente cada columna del conjunto de datos, se llegaron a las siguientes conclusiones:

* Columnas binarias (`'Partner'`, `'Dependents'`, `'PhoneService'` y `'PaperlessBilling'`) utilizan los valores `'Yes'` y `'No'`. Aunque podrían transformarse a booleanos (`True/False`) para facilitar operaciones lógicas o modelado, en este análisis se decidió mantener su formato original para favorecer la legibilidad.

* La columna `'SeniorCitizen'`, a diferencia de las anteriores, utiliza valores `0` y `1` de tipo `int64`. Para mantener un formato coherente con las demás variables binarias, se optó por **reemplazar 0 por** `'No'` **y 1 por**` 'Yes'`, y convertir la columna a tipo **string**.

* En la columna `'Churn'`, que indica si un cliente dejó o no la empresa, se detectaron tres valores únicos: `'Yes'`, `'No'` y `''` (cadena vacía). Dado que el valor vacío representa un estado indefinido, se decidió **eliminar estas filas para evitar inconsistencias analíticas**. El total de filas a eliminar es de **224**.

* Finalmente, la columna `'Charges.Total'`, que originalmente estaba en formato `object`, será convertida a `float64`, ya que contiene datos numéricos necesarios para cálculos posteriores.

---
---
* Ya que los datos están en inglés, se decidió hacer una modificación a los nombres de las columnas para mayor legibilidad.

| Nombre original       | Nuevo nombre       |
|------------------------|--------------------------|
| customerID             | ID_Cliente               |
| Churn                  | Abandono                 |
| gender                 | Genero                   |
| SeniorCitizen          | AdultoMayor              |
| Partner                | TienePareja              |
| Dependents             | Dependientes             |
| tenure                 | MesesAntiguedad          |
| PhoneService           | ServicioTelefonico       |
| MultipleLines          | LineasMultiples          |
| InternetService        | ServicioInternet         |
| OnlineSecurity         | SeguridadEnLinea         |
| OnlineBackup           | RespaldoEnLinea          |
| DeviceProtection       | ProteccionDispositivo    |
| TechSupport            | SoporteTecnico           |
| StreamingTV            | StreamingTV              |
| StreamingMovies        | StreamingPeliculas       |
| Contract               | TipoContrato             |
| PaperlessBilling       | FacturaElectronica       |
| PaymentMethod          | MetodoPago               |
| Charges.Monthly        | CargoMensual             |
| Charges.Total          | CargoTotal               |

---
---
**Datos Internos que también deben ser modificados para mayor legibilidad**

| Valor original              | Nuevo valor                   |
|-----------------------------|------------------------------------------|
| Yes                         | Sí                                       |
| No                          | No                                       |
| Female                      | Femenino                                 |
| Male                        | Masculino                                |
| No phone service            | Sin servicio telefónico                  |
| No internet service         | Sin servicio de internet                 |
| DSL                         | DSL                                      |
| Fiber optic                 | Fibra óptica                             |
| Month-to-month              | Mensual                                  |
| One year                    | Un año                                   |
| Two year                    | Dos años                                 |
| Mailed check                | Cheque por correo                        |
| Electronic check            | Cheque electrónico                       |
| Credit card (automatic)     | Tarjeta de crédito (automático)          |
| Bank transfer (automatic)   | Transferencia bancaria (automática)      |

## **Transformación de datos**

In [162]:
# Reemplaza 0 y 1 por 'No' y 'Yes' para mayor legibilidad y mantener un formato uniforme.
df['SeniorCitizen'] = df['SeniorCitizen'].map({0: 'No', 1: 'Yes'})

In [163]:
df.head()

customerID Churn  gender SeniorCitizen Partner Dependents  tenure  \
0  0002-ORFBO    No  Female            No     Yes        Yes       9   
1  0003-MKNFE    No    Male            No      No         No       9   
2  0004-TLHLJ   Yes    Male            No      No         No       4   
3  0011-IGKFF   Yes    Male           Yes     Yes         No      13   
4  0013-EXCHZ   Yes  Female           Yes     Yes         No       3   

  PhoneService MultipleLines InternetService  ... OnlineBackup  \
0          Yes            No             DSL  ...          Yes   
1          Yes           Yes             DSL  ...           No   
2          Yes            No     Fiber optic  ...           No   
3          Yes            No     Fiber optic  ...          Yes   
4          Yes            No     Fiber optic  ...           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No         Yes         Yes              No        One year   
1               No          No          No             Yes  Month-to-month   
2              Yes          No          No              No  Month-to-month   
3              Yes          No         Yes             Yes  Month-to-month   
4               No         Yes         Yes              No  Month-to-month   

  PaperlessBilling     PaymentMethod Charges.Monthly  Charges.Total  
0              Yes      Mailed check            65.6          593.3  
1               No      Mailed check            59.9          542.4  
2              Yes  Electronic check            73.9         280.85  
3              Yes  Electronic check            98.0        1237.85  
4              Yes      Mailed check            83.9          267.4  

[5 rows x 21 columns]

In [164]:
df.rename(columns={
    'customerID': 'ID_Cliente',
    'Churn': 'Abandono',
    'gender': 'Genero',
    'SeniorCitizen': 'AdultoMayor',
    'Partner': 'TienePareja',
    'Dependents': 'Dependientes',
    'tenure': 'MesesAntiguedad',
    'PhoneService': 'ServicioTelefonico',
    'MultipleLines': 'LineasMultiples',
    'InternetService': 'ServicioInternet',
    'OnlineSecurity': 'SeguridadEnLinea',
    'OnlineBackup': 'RespaldoEnLinea',
    'DeviceProtection': 'ProteccionDispositivo',
    'TechSupport': 'SoporteTecnico',
    'StreamingTV': 'StreamingTV',
    'StreamingMovies': 'StreamingPeliculas',
    'Contract': 'TipoContrato',
    'PaperlessBilling': 'FacturaElectronica',
    'PaymentMethod': 'MetodoPago',
    'Charges.Monthly': 'CargoMensual',
    'Charges.Total': 'CargoTotal'
}, inplace=True)

In [170]:
# Diccionario con las traducciones
valores_a_reemplazar = {
    'Yes': 'Sí',
    'No': 'No',
    '': 'Desconocido',
    'Female': 'Femenino',
    'Male': 'Masculino',
    'No phone service': 'Sin servicio telefónico',
    'No internet service': 'Sin servicio de internet',
    'DSL': 'DSL',
    'Fiber optic': 'Fibra óptica',
    'Month-to-month': 'Mensual',
    'One year': 'Un año',
    'Two year': 'Dos años',
    'Mailed check': 'Cheque por correo',
    'Electronic check': 'Cheque electrónico',
    'Credit card (automatic)': 'Tarjeta de crédito (automático)',
    'Bank transfer (automatic)': 'Transferencia bancaria (automática)'
}

# Aplicamos los reemplazos a todo el DataFrame
df.replace(valores_a_reemplazar, inplace=True)

In [167]:
df.head()

ID_Cliente Abandono     Genero AdultoMayor TienePareja Dependientes  \
0  0002-ORFBO       No   Femenino          No          Sí           Sí   
1  0003-MKNFE       No  Masculino          No          No           No   
2  0004-TLHLJ       Sí  Masculino          No          No           No   
3  0011-IGKFF       Sí  Masculino          Sí          Sí           No   
4  0013-EXCHZ       Sí   Femenino          Sí          Sí           No   

   MesesAntiguedad ServicioTelefonico LineasMultiples ServicioInternet  ...  \
0                9                 Sí              No              DSL  ...   
1                9                 Sí              Sí              DSL  ...   
2                4                 Sí              No     Fibra óptica  ...   
3               13                 Sí              No     Fibra óptica  ...   
4                3                 Sí              No     Fibra óptica  ...   

  RespaldoEnLinea ProteccionDispositivo SoporteTecnico StreamingTV  \
0              Sí                    No             Sí          Sí   
1              No                    No             No          No   
2              No                    Sí             No          No   
3              Sí                    Sí             No          Sí   
4              No                    No             Sí          Sí   

  StreamingPeliculas TipoContrato FacturaElectronica          MetodoPago  \
0                 No       Un año                 Sí   Cheque por correo   
1                 Sí      Mensual                 No   Cheque por correo   
2                 No      Mensual                 Sí  Cheque electrónico   
3                 Sí      Mensual                 Sí  Cheque electrónico   
4                 No      Mensual                 Sí   Cheque por correo   

  CargoMensual  CargoTotal  
0         65.6       593.3  
1         59.9       542.4  
2         73.9      280.85  
3         98.0     1237.85  
4         83.9       267.4  

[5 rows x 21 columns]

In [168]:
# Revisión de datos únicos para cada columna
for col in df.columns:
    if df[col].dtype == 'object':
        print(f"Valores únicos en '{col}':")
        print(df[col].unique())
        print("—" * 40)

Valores únicos en 'ID_Cliente':
['0002-ORFBO' '0003-MKNFE' '0004-TLHLJ' ... '9992-UJOEL' '9993-LHIEB'
 '9995-HOTOH']
————————————————————————————————————————
Valores únicos en 'Abandono':
['No' 'Sí' None]
————————————————————————————————————————
Valores únicos en 'Genero':
['Femenino' 'Masculino']
————————————————————————————————————————
Valores únicos en 'AdultoMayor':
['No' 'Sí']
————————————————————————————————————————
Valores únicos en 'TienePareja':
['Sí' 'No']
————————————————————————————————————————
Valores únicos en 'Dependientes':
['Sí' 'No']
————————————————————————————————————————
Valores únicos en 'ServicioTelefonico':
['Sí' 'No']
————————————————————————————————————————
Valores únicos en 'LineasMultiples':
['No' 'Sí' 'Sin servicio telefónico']
————————————————————————————————————————
Valores únicos en 'ServicioInternet':
['DSL' 'Fibra óptica' 'No']
————————————————————————————————————————
Valores únicos en 'SeguridadEnLinea':
['No' 'Sí' 'Sin servicio de internet']
————————

## **Carga y análisis**

## **Informe final**